File: wtt_module/customization/custom/supplier_quotation.py
# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
# License: GNU General Public License v3. See license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import flt, nowdate, add_days
from frappe.model.mapper import get_mapped_doc

from erpnext.controllers.buying_controller import BuyingController
from erpnext.buying.utils import validate_for_items
import json
import csv
import requests
form_grid_templates = {
	"items": "templates/form_grid/item_grid.html"
}

class SupplierQuotation(BuyingController):
	def validate(self):
		# # for rfq supplier submit only individual table append but comparison need combined table
		# if(self.combined_table==[] and self.supplier_data_table==[]):
		# 	for i in frappe.db.sql("SELECT item_code,sum(qty)as qty,uom,description,technical_description,rate,sum(amount)as amount,sum(price_list_rate)as price_list_rate,discount_percentage,discount_amount,item_tax_template,material_request FROM `tabSupplier Quotation Item` WHERE parent='"+str(self.name)+"' GROUP BY item_code ORDER BY technical_description",as_dict=1):
		# 		self.append("supplier_data_table",i)
		# 	for j in frappe.db.sql("SELECT item_code,sum(qty)as qty,uom,description,technical_description,rate as rate2,sum(amount)as amount2,sum(price_list_rate)as price_list_rate2,discount_percentage,discount_amount,item_tax_template,material_request FROM `tabSupplier Quotation Item` WHERE parent='"+str(self.name)+"' GROUP BY item_code ORDER BY technical_description",as_dict=1):
		# 		self.append("combined_table",j)

		super(SupplierQuotation, self).validate()
		if not self.status:
			self.status = "Draft"

		from erpnext.controllers.status_updater import validate_status
		validate_status(self.status, ["Draft", "Submitted", "Stopped",
			"Cancelled"])

		validate_for_items(self)
		self.validate_with_previous_doc()
		self.validate_uom_is_integer("uom", "qty")

	def on_submit(self):
		# for i in self.get("combined_table"):
		# 	if(i.approval=='Approved'):
		# 		frappe.db.sql("UPDATE `tabSupplier Quotation Item` SET quoted_qty='"+str(i.qty)+"',approval='"+str(i.approval)+"' WHERE item_code='"+str(i.item_code)+"' and parent='"+str(self.name)+"' ",as_dict=1)		
		frappe.db.set(self, "status", "Submitted")
		self.update_rfq_supplier_status(1)

	def on_cancel(self):
		frappe.db.set(self, "status", "Cancelled")
		self.update_rfq_supplier_status(0)

	def validate_with_previous_doc(self):
		super(SupplierQuotation, self).validate_with_previous_doc({
			"Material Request": {
				"ref_dn_field": "prevdoc_docname",
				"compare_fields": [["company", "="]],
			},
			"Material Request Item": {
				"ref_dn_field": "prevdoc_detail_docname",
				"compare_fields": [["item_code", "="], ["uom", "="]],
				"is_child_table": True
			}
		})
	def update_rfq_supplier_status(self, include_me):
		rfq_list = set([])
		for item in self.items:
			if item.request_for_quotation:
				rfq_list.add(item.request_for_quotation)
		for rfq in rfq_list:
			doc = frappe.get_doc('Request for Quotation', rfq)
			doc_sup = frappe.get_all('Request for Quotation Supplier', filters=
				{'parent': doc.name, 'supplier': self.supplier}, fields=['name', 'quote_status'])

			doc_sup = doc_sup[0] if doc_sup else None
			if not doc_sup:
				frappe.throw(_("Supplier {0} not found in {1}").format(self.supplier,
					"<a href='desk#Form/Request for Quotation/{0}'> Request for Quotation {0} </a>".format(doc.name)))

			quote_status = _('Received')
			for item in doc.items:
				sqi_count = frappe.db.sql("""
					SELECT
						COUNT(sqi.name) as count
					FROM
						`tabSupplier Quotation Item` as sqi,
						`tabSupplier Quotation` as sq
					WHERE sq.supplier = %(supplier)s
						AND sqi.docstatus = 1
						AND sq.name != %(me)s
						AND sqi.request_for_quotation_item = %(rqi)s
						AND sqi.parent = sq.name""",
					{"supplier": self.supplier, "rqi": item.name, 'me': self.name}, as_dict=1)[0]
				self_count = sum(my_item.request_for_quotation_item == item.name
					for my_item in self.items) if include_me else 0
				if (sqi_count.count + self_count) == 0:
					quote_status = _('Pending')
			if quote_status == _('Received') and doc_sup.quote_status == _('No Quote'):
				frappe.msgprint(_("{0} indicates that {1} will not provide a quotation, but all items \
					have been quoted. Updating the RFQ quote status.").format(doc.name, self.supplier))
				frappe.db.set_value('Request for Quotation Supplier', doc_sup.name, 'quote_status', quote_status)
				frappe.db.set_value('Request for Quotation Supplier', doc_sup.name, 'no_quote', 0)
			elif doc_sup.quote_status != _('No Quote'):
				frappe.db.set_value('Request for Quotation Supplier', doc_sup.name, 'quote_status', quote_status)

def get_list_context(context=None):
	from erpnext.controllers.website_list_for_contact import get_list_context
	list_context = get_list_context(context)
	list_context.update({
		'show_sidebar': True,
		'show_search': True,
		'no_breadcrumbs': True,
		'title': _('Supplier Quotation'),
	})

	return list_context

@frappe.whitelist()
def make_purchase_order(source_name, target_doc=None):
	def set_missing_values(source, target):
		target.ignore_pricing_rule = 1
		target.run_method("set_missing_values")
		target.run_method("get_schedule_dates")
		target.run_method("calculate_taxes_and_totals")

	def update_item(obj, target, source_parent):
		target.stock_qty = flt(obj.qty) * flt(obj.conversion_factor)

	doclist = get_mapped_doc("Supplier Quotation", source_name,		{
		"Supplier Quotation": {
			"doctype": "Purchase Order",
			"validation": {
				"docstatus": ["=", 1],
			}
		},
		"Supplier Quotation Item": {
			"doctype": "Purchase Order Item",
			"field_map": [
				["name", "supplier_quotation_item"],
				["parent", "supplier_quotation"],
				["material_request", "material_request"],
				["material_request_item", "material_request_item"],
				["sales_order", "sales_order"]
			],
			"postprocess": update_item,
			"condition": lambda doc: doc.approval=="Approved" and doc.qty>0
		},
		"SQ Combined Table": {
			"doctype": "PO Combined Table",
			"field_map": [
				["price_list_rate2","price_list_rate"],
				["discount_percentage2","discount_percentage"],
				["discount_amount2","discount_amount"],
				["rate2","rate1"],
				["amount2","amount1"],
				["name", "sq_combined_item"],
				["parent", "supplier_quotation"],
				["material_request", "material_request"],
				["material_request_item", "material_request_item"],
				["sales_order", "sales_order"]
			],
			"postprocess": update_item,
			"condition": lambda doc: doc.approval=="Approved" and doc.qty>0
		},
		"Purchase Taxes and Charges": {
			"doctype": "Purchase Taxes and Charges",
		}
	}, target_doc, set_missing_values)

	return doclist

@frappe.whitelist()
def make_quotation(source_name, target_doc=None):
	doclist = get_mapped_doc("Supplier Quotation", source_name, {
		"Supplier Quotation": {
			"doctype": "Quotation",
			"field_map": {
				"name": "supplier_quotation",
			}
		},
		"Supplier Quotation Item": {
			"doctype": "Quotation Item",
			"condition": lambda doc: frappe.db.get_value("Item", doc.item_code, "is_sales_item")==1,
			"add_if_empty": True
		}
	}, target_doc)

	return doclist


@frappe.whitelist()
def set_rate(name):
	doc=frappe.get_doc("Supplier Quotation",name)
	for i in doc.combined_table:
		for sub in frappe.db.sql("SELECT name,qty from `tabSupplier Quotation Item` WHERE item_code='"+str(i.item_code)+"' ",as_dict=1):
			frappe.db.sql("UPDATE `tabSupplier Quotation Item` set price_list_rate='"+str(i.price_list_rate2)+"',discount_percentage='"+str(i.discount_percentage2)+"',discount_amount='"+str(i.discount_amount2)+"',rate='"+str(i.rate2)+"',amount='"+str(i.rate2*sub.qty)+"',supplier_description='"+str(i.supplier_description)+"' WHERE parent='"+str(name)+"' and item_code='"+str(i.item_code)+"' and name='"+str(sub.name)+"' ")
		
	return doc

@frappe.whitelist()
def get_data_from_excel(file_path):
	frappe.msgprint(str(file_path))
	flname = str(file_path)
	arr=[]
	val=[]
	v=[]
	with open(flname) as file_obj:
	    reader_obj = csv.reader(file_obj)
	    for row in reader_obj:
	        arr.append(row)

	    for j in range(len(arr)):
	        vals = {}
	        for k in range(len(arr[j])):
	            vals["col"+str(k)]=arr[j][k]
	        val.append(vals)
	frappe.msgprint(str(val))

File: wtt_module/customization/custom/supplier_quotation.js
{% include 'erpnext/buying/doctype/supplier_quotation/supplier_quotation.js' %};
frappe.ui.form.on("Supplier Quotation", {
	setup: function(frm) {
		frm.get_docfield('combined_table').allow_bulk_edit = 1;
		// frm.get_docfield('supplier_data_table').allow_bulk_edit = 1;
		
	},
	upload:function(frm){
		frappe.call({
			method:"wtt_module.customization.custom.supplier_quotation.get_data_from_excel",
			args:{
				file_path:"C:/Users/erp-1/Downloads/RFQ-22-00168.csv"
			}
		})
	},
	get_combined:function(frm){
				
	},
	custom_duty_percentage:function(frm){
		frm.set_value("custom_duty_amount",frm.doc.base_total*(frm.doc.custom_duty_percentage/100));
		frm.refresh_field("custom_duty_amount");
	},
	import_duty_percentage:function(frm){
		frm.set_value("import_duty_amount",frm.doc.base_total*(frm.doc.import_duty_percentage/100));
		frm.refresh_field("import_duty_amount");
	},
	validate:function(frm,cdt,cdn){
		frm.set_value("import_and_custom",(frm.doc.import_duty_amount+frm.doc.custom_duty_amount));
		
		if(frm.doc.conversion_rate>0){
		$.each(frm.doc.supplier_data_table, function (i, cd) {
			cd.rate_inr = (cd.rate*frm.doc.conversion_rate)
			cd.amount_inr = (cd.amount*frm.doc.conversion_rate)
		})
		$.each(frm.doc.combined_table, function (i, cd) {
			cd.rate_inr = (cd.rate2*frm.doc.conversion_rate)
			cd.amount_inr = (cd.amount2*frm.doc.conversion_rate)
		})
		}
	},
	before_save:function(frm){		
		var r1=frm.doc.combined_table
		if(r1.length == 0){
			frm.clear_table("combined_table");
			frm.refresh_field("combined_table");
			var array=[]
			$.each(frm.doc.supplier_data_table, function (index, vv) {
		        	array.push({
		        		"item_code":vv.item_code,
		        		"description":vv.description,
		        		"item_name":vv.item_name,
		        		"technical_description":vv.technical_description,
		        		"supplier_description":vv.supplier_description,
		        		"mr_qty":vv.mr_qty,
		        		"qty":vv.qty,
		        		"rate":vv.rate,
		        		"amount":vv.amount,
		        		"item_tax_template":vv.item_tax_template,
		        		"uom":vv.uom,
		        		"stock_uom":vv.stock_uom,
		        		"conversion_rate":vv.conversion_rate,
		        		"price_list_rate":vv.price_list_rate,
		        		"discount_percentage":vv.discount_percentage,
		        		"discount_amount":vv.discount_amount
		        	})
				});
			// alert(array)
			for(var i in array){
			var child = frm.add_child("combined_table");
			frappe.model.set_value(child.doctype, child.name, "item_code", array[i].item_code);
			frappe.model.set_value(child.doctype, child.name, "description", array[i].description);
			frappe.model.set_value(child.doctype, child.name, "technical_description", array[i].technical_description);
			frappe.model.set_value(child.doctype, child.name, "item_name", array[i].item_name);
			frappe.model.set_value(child.doctype, child.name, "supplier_description", array[i].supplier_description);
			frappe.model.set_value(child.doctype, child.name, "mr_qty",array[i].mr_qty);
			frappe.model.set_value(child.doctype, child.name, "qty", array[i].qty);
			frappe.model.set_value(child.doctype, child.name, "price_list_rate2", array[i].price_list_rate);
			frappe.model.set_value(child.doctype, child.name, "discount_percentage2", array[i].discount_percentage);
			frappe.model.set_value(child.doctype, child.name, "discount_amount2", array[i].discount_amount);
			frappe.model.set_value(child.doctype, child.name, "rate2", array[i].rate);
			frappe.model.set_value(child.doctype, child.name, "amount2", array[i].rate * array[i].qty);
			frappe.model.set_value(child.doctype, child.name, "item_tax_template", array[i].item_tax_template);
			frappe.model.set_value(child.doctype, child.name, "uom", array[i].uom);
			frappe.model.set_value(child.doctype, child.name, "stock_uom", array[i].stock_uom);
			frappe.model.set_value(child.doctype, child.name, "conversion_rate", array[i].conversion_rate);
			}
			frm.refresh_field("combined_table");
			
		}
		
	},
	after_save:function(frm,cdt,cdn){
		frappe.call({
			method:"wtt_module.customization.custom.supplier_quotation.set_rate",
			args:{
				name:frm.doc.name
			},
			callback(r){
				frm.reload_doc()
				frm.save()
			}
		})
	},
	// btn1:function(frm,cdt,cdn){
	// 	var htvalue='<style>th{font-size: 15px;text-align:center;color:#6A5ACD;}</style>'
	// 	htvalue+="<table border='1px' style='margin-top:-1px;width:100%'><tr><th style='width:1%;'>S.NO</th><th style='width:2%'>DESCRIPTION</th><th style='width:15%'>TECHNICAL DESCRIPTION</th><th style='text-align:center;width:5%'>MR QTY</th><th style='text-align:center;width:5%'>QTY</th><th style='width:5%'>PRICE RATE</th><th style='width:5%'>DISCOUNT</th><th style='width:5%'>RATE</th><th style='width:5%'>AMOUNT</th></tr>"
	// 	$.each(frm.doc.supplier_data_table || [], function(i, v) {
	// 		var result=''
	// 		var re=''
			
	// 		if(v.technical_description)
	// 		{
	// 		var go=[]
	// 		var val=v.technical_description.split(",")
	// 		var gg='<br>'
	// 		for(var g in val)
	// 		{
	// 			go.push(val[g])
	// 			go.push(gg)
	// 		}
	// 		result = go.toString().replace(/,/g, "");
	// 		}
	// 		if(v.remarks)
	// 		{
	// 			re=v.remarks
	// 		}
	// 		htvalue+='<tr style="text-align:center;"><td align="center">'+v.idx+'.</td><td align="center">'+v.description+'<br></td><td align="left">'+result+'</td><td>'+v.mr_qty+'</td><td>'+v.qty+'</td><td>'+v.price_list_rate+'</td><td>'+v.discount_percentage+'</td><td>'+v.rate+'</td><td>'+v.amount+'</td></tr>'
	// 	});
	// 	$(frm.fields_dict['tab2'].wrapper).html("")
	// 	$(frm.fields_dict['tab3'].wrapper).html("")
	// 	$(frm.fields_dict['tab1'].wrapper).html(htvalue);
	// },
	// btn2:function(frm,cdt,cdn){
	// 	var htvalue='<style>th{font-size: 15px;text-align:center;color:#6A5ACD;}</style>'
	// 	htvalue+="<table border='1px' style='margin-top:-1px;width:100%'><tr><th style='width:1%;'>S.NO</th><th style='width:2%'>DESCRIPTION</th><th style='width:15%'>TECHNICAL DESCRIPTION</th><th style='text-align:center;width:5%'>MR QTY</th><th style='text-align:center;width:5%'>QUOTED QTY</th><th style='width:5%'>PRICE RATE</th><th style='width:5%'>DISCOUNT</th><th style='width:5%'>RATE</th><th style='width:5%'>AMOUNT</th></tr>"
	// 	$.each(frm.doc.combined_table || [], function(i, v) {
	// 		var result=''
	// 		var re=''
			
	// 		if(v.technical_description)
	// 		{
	// 		var go=[]
	// 		var val=v.technical_description.split(",")
	// 		var gg='<br>'
	// 		for(var g in val)
	// 		{
	// 			go.push(val[g])
	// 			go.push(gg)
	// 		}
	// 		result = go.toString().replace(/,/g, "");
	// 		}
	// 		if(v.remarks)
	// 		{
	// 			re=v.remarks
	// 		}
	// 		htvalue+='<tr style="text-align:center;"><td align="center">'+v.idx+'.</td><td align="center">'+v.description+'<br></td><td align="left">'+result+'</td><td>'+v.mr_qty+'</td><td>'+v.qty+'</td><td>'+v.price_list_rate2+'</td><td>'+v.discount_percentage2+'</td><td>'+v.rate2+'</td><td>'+v.amount2+'</td></tr>'
	// 	});
	// 	$(frm.fields_dict['tab1'].wrapper).html("")
	// 	$(frm.fields_dict['tab3'].wrapper).html("")
	// 	$(frm.fields_dict['tab2'].wrapper).html(htvalue);
	// },
	// btn3:function(frm,cdt,cdn){
	// 	var htvalue='<style>th{font-size: 15px;text-align:center;color:#6A5ACD;}</style>'
	// 	htvalue+="<table border='1px' style='margin-top:-1px;width:100%'><tr><th style='width:1%;'>S.NO</th><th style='width:2%'>DESCRIPTION</th><th style='width:15%'>TECHNICAL DESCRIPTION</th><th style='text-align:center;width:5%'>MR QTY</th><th style='text-align:center;width:5%'>REQUIRED QTY</th><th style='text-align:center;width:5%'>APPROVED QTY</th><th style='width:5%'>PRICE RATE</th><th style='width:5%'>DISCOUNT</th><th style='width:5%'>RATE</th><th style='width:5%'>AMOUNT</th></tr>"
	// 	$.each(frm.doc.items || [], function(i, v) {
	// 		var result=''
	// 		var re=''
			
	// 		if(v.technical_description)
	// 		{
	// 		var go=[]
	// 		var val=v.technical_description.split(",")
	// 		var gg='<br>'
	// 		for(var g in val)
	// 		{
	// 			go.push(val[g])
	// 			go.push(gg)
	// 		}
	// 		result = go.toString().replace(/,/g, "");
	// 		}
	// 		if(v.remarks)
	// 		{
	// 			re=v.remarks
	// 		}
	// 		htvalue+='<tr style="text-align:center;"><td align="center">'+v.idx+'.</td><td align="center">'+v.description+'<br></td><td align="left">'+result+'</td><td>'+v.mr_qty+'</td><td>'+v.qty+'</td><td>'+v.quoted_qty+'</td><td>'+v.price_list_rate+'</td><td>'+v.discount_percentage+'</td><td>'+v.rate+'</td><td>'+v.amount+'</td></tr>'
	// 	});
	// 	$(frm.fields_dict['tab2'].wrapper).html("")
	// 	$(frm.fields_dict['tab1'].wrapper).html("")
	// 	$(frm.fields_dict['tab3'].wrapper).html(htvalue);
	// },
	// refresh:function(frm,cdt,cdn){
	// 	$(frm.fields_dict['tab2'].wrapper).html("")
	// 	$(frm.fields_dict['tab1'].wrapper).html("")
	// 	$(frm.fields_dict['tab3'].wrapper).html("");

	// }
});

frappe.ui.form.on("Supplier Data Table", {
	rate: function(frm,cdt, cdn){
		var cd = locals[cdt][cdn];
		frappe.model.set_value(cdt, cdn, "amount", (cd.qty*cd.rate));

	},
	qty:function(frm,cdt,cdn){
		var cd = locals[cdt][cdn];
		frappe.model.set_value(cdt, cdn, "amount", (cd.qty*cd.rate));
	},
	amount:function(frm,cdt,cdn){
		calculate_time_and_amount_table1(frm);
	}
});

frappe.ui.form.on("SQ Combined Table", {
	refresh:function(frm){
		const collection = document.getElementsByClassName(".col.grid-static-col.col-xs-3");
		for (let i = 0; i < collection.length; i++) {
		  collection[i].style.height = "fit-content";
		}
	},
	discount_percentage2:function(frm,cdt,cdn){
		var cd = locals[cdt][cdn];
		var amt=(cd.discount_percentage2/100)*cd.price_list_rate2
		if(cd.descount_percentage2>100){
		frappe.model.set_value(cdt, cdn, "discount_percentage2",100)
		}
		frappe.model.set_value(cdt, cdn, "discount_amount2",amt);
		frappe.model.set_value(cdt,cdn,"rate2",cd.price_list_rate2-amt)
		frappe.model.set_value(cdt,cdn,"amount2",cd.qty*(cd.price_list_rate2-amt))
	},
	price_list_rate2:function(frm,cdt,cdn){
		var cd = locals[cdt][cdn];
		var amt=(cd.discount_percentage2/100)*cd.price_list_rate2
		if(cd.descount_percentage2>100){
		frappe.model.set_value(cdt, cdn, "discount_percentage2",100)
		}
		frappe.model.set_value(cdt, cdn, "discount_amount2",amt);
		frappe.model.set_value(cdt,cdn,"rate2",cd.price_list_rate2-amt)
		frappe.model.set_value(cdt,cdn,"amount2",cd.qty*(cd.price_list_rate2-amt))
		if(frm.doc.conversion_rate>0){
			frappe.model.set_value(cdt, cdn, "rate_inr", ((cd.price_list_rate2-amt)*frm.doc.conversion_rate));
			frappe.model.set_value(cdt, cdn, "amount_inr", (((cd.price_list_rate2-amt)*frm.doc.conversion_rate)*cd.qty));
		}
	},

	rate2: function(frm,cdt, cdn){
		var cd = locals[cdt][cdn];
		frappe.model.set_value(cdt, cdn, "amount2", (cd.qty*cd.rate2));
		if(frm.doc.conversion_rate>0){
			frappe.model.set_value(cdt, cdn, "rate_inr", (cd.rate2*frm.doc.conversion_rate));
			frappe.model.set_value(cdt, cdn, "amount_inr", ((cd.rate2*frm.doc.conversion_rate)*cd.qty));
		}
		
	},
	qty:function(frm,cdt,cdn){
		var cd = locals[cdt][cdn];
		frappe.model.set_value(cdt, cdn, "amount2", (cd.qty*cd.rate2));
		frappe.model.set_value(cdt, cdn, "amount_inr", (cd.qty*cd.rate2)*frm.doc.conversion_rate);
	},
	amount2:function(frm,cdt,cdn){
		calculate_time_and_amount(frm);
	}
});

frappe.ui.form.on("Supplier Quotation Item", {
	rate:function(frm,cdt,cdn){
		var cd = locals[cdt][cdn];
		if(frm.doc.conversion_rate){
			frappe.model.set_value(cdt, cdn, "rate_inr", (cd.rate2*frm.doc.conversion_rate));
			frappe.model.set_value(cdt, cdn, "amount_inr", ((cd.rate2*frm.doc.conversion_rate)*cd.qty));
		}
		// else
	},
	uom_rate: function(frm,cdt, cdn){
		calculate_total(frm, cdt, cdn);
	}
});

var calculate_total = function(frm, cdt, cdn) {
	var child = locals[cdt][cdn];
	frappe.model.set_value(cdt, cdn, "rate", ((child.stock_qty/child.qty)*child.uom_rate));
	}
var calculate_time_and_amount = function(frm) {
var temp = frm.doc.combined_table;
var i,sum=0
for(i=0;i<temp.length;i++)
{
sum+=temp[i].amount2;
}
frm.set_value("supplier_total",sum);
};

var calculate_time_and_amount_table1 = function(frm) {
var temp = frm.doc.supplier_data_table;
var i,sum=0
for(i=0;i<temp.length;i++)
{
sum+=temp[i].amount;
}
frm.set_value("supplier_total2",sum);
};